### PR TITLE
Sync location fixtures when needed

### PR DESCRIPTION
**Pull request by proteusvacuum**
This adds a `UserFixtureStatus` model which can be used to invalidate fixtures for a particular user. It is currently only implemented for `Location`. 

It also adds a `last_modified` column to `LocationTypes` which is also used to re-sync fixtures whenever a location_type is changed.

I think, unfortunately that this introduces a bug that I'm not sure how to fix (!)
If you add someone to one location, sync, remove them from that location and add them to a different location and sync, everything works fine.
However, if you add them to a location, sync, then remove them from that location and sync again (i.e. that user is not associated with a location), you end up with a `BadStateException: Phone state hash mismatch` when trying to sync. The user also maintains access to the old location they had access to on the phone.

http://manage.dimagi.com/default.asp?174551

@snopoke @czue @esoergel 

cc: @biyeun

### CODE REVIEW

**Comment by czue** [comment]
@proteusvacuum I don't think the bad state exception bug is introduced by this - i think it also exists on prod (though i hadn't thought of it until you mentioned)

File: corehq/apps/locations/models.py [hunk 1]
```diff
@@ -77,6 +77,7 @@ class LocationType(models.Model):
     administrative = models.BooleanField(default=False)
     shares_cases = models.BooleanField(default=False)
     view_descendants = models.BooleanField(default=False)
+    last_modified = models.DateTimeField(auto_now=True)
```
Comment (czue) [inline corehq/apps/locations/models.py]: my read of https://docs.djangoproject.com/en/1.8/ref/models/fields/#datefield is that this uses the project timezone and not UTC. i was going to wonder if we should change this but i see we use it everywhere in the code so seems ok

Comment (czue) [inline corehq/apps/locations/models.py]: (i think we set the project timezone to UTC so practically it shouldn't matter)

File: corehq/apps/locations/tests/test_location_fixtures.py [hunk 2]
```diff
@@ -1,28 +1,97 @@
...
+from corehq.apps.fixtures.models import UserFixtureStatus, UserFixtureType
 
 
-class LocationFixturesTest(SimpleTestCase):
-
-    def test_metadata(self):
-        state = LocationType(
-            domain="test-domain",
+class LocationFixturesTest(TestCase):
+    def setUp(self):
```
Comment (czue) [inline corehq/apps/locations/tests/test_location_fixtures.py]: if you don't need these cleared between tests you can use `setUpClass` and `tearDownClass` which perform a bit better

File: corehq/apps/locations/tests/test_location_fixtures.py [hunk 3]
```diff
@@ -1,28 +1,97 @@
...
+        fixture = _location_to_fixture(location_db, location, self.location_type)
         location_data = {
             e.tag: e.text for e in fixture.find('location_data')
         }
         self.assertEquals(location_data, location.metadata)
+
+    def test_should_sync_locations_change_location_type(self):
+        """
+        When location_type gets changed, we should resync locations
+        Had to force update last_modified because it has the auto_now
```
Comment (czue) [inline corehq/apps/locations/tests/test_location_fixtures.py]: super nit picky but the second half of this doc string is more like a comment

**Comment by czue** [comment]
build fail looks real. i think it will be fixed if you fix [this migration file import](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/locations/migrations/0013_get_stock_levels_from_config.py#L3) to use the south ORM instead of the default (which it should have been doing all along)

```
Traceback (most recent call last):
  File "manage.py", line 65, in <module>
    execute_from_command_line(sys.argv)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 399, in execute_from_command_line
    utility.execute()
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 392, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/commands/test.py", line 50, in run_from_argv
    super(Command, self).run_from_argv(argv)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/base.py", line 242, in run_from_argv
    self.execute(*args, **options.__dict__)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/commands/test.py", line 71, in execute
    super(Command, self).execute(*args, **options)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/base.py", line 285, in execute
    output = self.handle(*args, **options)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/south/management/commands/test.py", line 8, in handle
    super(Command, self).handle(*args, **kwargs)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/commands/test.py", line 88, in handle
    failures = test_runner.run_tests(test_labels)
  File "/home/travis/build/dimagi/commcare-hq/testrunner.py", line 233, in run_tests
    failures += self.run_db_tests(db_suite)
  File "/home/travis/build/dimagi/commcare-hq/testrunner.py", line 22, in _inner
    return fn(*args, **kwargs)
  File "/home/travis/build/dimagi/commcare-hq/testrunner.py", line 206, in run_db_tests
    old_config = self.setup_databases()
  File "/home/travis/build/dimagi/commcare-hq/testrunner.py", line 73, in setup_databases
    return super(HqTestSuiteRunner, self).setup_databases(**kwargs)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/couchdbkit/ext/django/testrunner.py", line 64, in setup_databases
    return super(CouchDbKitTestSuiteRunner, self).setup_databases(**kwargs)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/test/runner.py", line 107, in setup_databases
    return setup_databases(self.verbosity, self.interactive, **kwargs)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/test/runner.py", line 279, in setup_databases
    verbosity, autoclobber=not interactive)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/south/hacks/django_1_0.py", line 103, in wrapper
    return f(*args, **kwargs)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/db/backends/creation.py", line 339, in create_test_db
    load_initial_data=False)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 159, in call_command
    return klass.execute(*args, **defaults)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/base.py", line 285, in execute
    output = self.handle(*args, **options)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/base.py", line 415, in handle
    return self.handle_noargs(**options)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/south/management/commands/syncdb.py", line 103, in handle_noargs
    management.call_command('migrate', **options)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 159, in call_command
    return klass.execute(*args, **defaults)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/base.py", line 285, in execute
    output = self.handle(*args, **options)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/south/management/commands/migrate.py", line 111, in handle
    ignore_ghosts = ignore_ghosts,
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/south/migration/__init__.py", line 220, in migrate_app
    success = migrator.migrate_many(target, workplan, database)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/south/migration/migrators.py", line 331, in migrate_many
    result = self.migrate(migration, database)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/south/migration/migrators.py", line 133, in migrate
    result = self.run(migration, database)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/south/migration/migrators.py", line 114, in run
    return self.run_migration(migration, database)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/south/migration/migrators.py", line 84, in run_migration
    migration_function()
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/south/migration/migrators.py", line 60, in <lambda>
    return (lambda: direction(orm))
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/locations/migrations/0013_get_stock_levels_from_config.py", line 10, in forwards
    for loc_type in LocationType.objects.iterator():
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/db/models/query.py", line 220, in iterator
    for row in compiler.results_iter():
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 713, in results_iter
    for rows in self.execute_sql(MULTI):
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 786, in execute_sql
    cursor.execute(sql, params)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/db/backends/util.py", line 53, in execute
    return self.cursor.execute(sql, params)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/db/utils.py", line 99, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/db/backends/util.py", line 53, in execute
    return self.cursor.execute(sql, params)
ProgrammingError: column locations_locationtype.last_modified does not exist
LINE 1: ...es", "locations_locationtype"."view_descendants", "locations...
```

**Comment by czue** [comment]
this looks good to me - I think fine to merge as soon as tests pass and you confirm the phone state thing is already happening on prod - though my instinct is that there's no way this change could have introduced that since it would be caused by cases not fixtures.

File: corehq/apps/fixtures/models.py [hunk 4]
```diff
@@ -522,3 +523,22 @@ def for_all_item_ids(cls, item_ids, domain):
...
+        (LOCATION, "Location")
+    )
+
+
+class UserFixtureStatus(models.Model):
+    """
+    Keeps track of when a user needs to re-sync a fixture
+    """
+    user_id = models.CharField(max_length=100, db_index=True)
+    fixture_type = models.CharField(max_length=50)
```
Comment (snopoke) [inline corehq/apps/fixtures/models.py]: Could make this an int field and use less space

Comment (snopoke) [inline corehq/apps/fixtures/models.py]: Or even a small int since we're not going to have that many fixtures

File: corehq/apps/locations/fixtures.py [hunk 5]
```diff
@@ -27,16 +29,35 @@ def __contains__(self, item):
...
-def should_sync_locations(last_sync, location_db):
+def fixture_last_modified(user):
+    """
+    Return when the fixture was last modified
+    """
+    try:
+        return UserFixtureStatus.objects.get(
+            user_id=user._id, fixture_type=UserFixtureType.LOCATION).last_modified
+    except UserFixtureStatus.DoesNotExist:
+        return datetime(1970, 1, 1)
```
Comment (snopoke) [inline corehq/apps/locations/fixtures.py]: Not problem with this but could also use `datetime.min`

File: corehq/apps/locations/fixtures.py [hunk 6]
```diff
@@ -27,16 +29,35 @@ def __contains__(self, item):
         return item in self.by_id
 
 
-def should_sync_locations(last_sync, location_db):
+def fixture_last_modified(user):
```
Comment (snopoke) [inline corehq/apps/locations/fixtures.py]: Perhaps this doesn't make sense right now but if we add more fixtures to this table then it would be better for this function to get all their last modified dates in one query and return and object.

Should this get cached at all?

**Comment by snopoke** [comment]
Nothing blocking from me :white_check_mark:

File: corehq/apps/locations/models.py (hunk 1 above)
Comment (esoergel) [inline corehq/apps/locations/models.py]: Seems reasonable to ping @dannyroberts just in case.  Danny, any objections to using `models.DateTimeField(auto_now=True)`?  Does it attach tz information or is it safe to store with the assumption that it's always UTC?

File: corehq/apps/fixtures/models.py (hunk 4 above)
Comment (esoergel) [inline corehq/apps/fixtures/models.py]: You could also add a choices option to this field https://docs.djangoproject.com/en/1.6/ref/models/fields/#choices although reading the docs, it doesn't look like it actually does any validation around that, unless this ends up with a django admin view or similar.  Yeah, probably fine without.

File: corehq/apps/locations/fixtures.py (hunk 6 above)
Comment (esoergel) [inline corehq/apps/locations/fixtures.py]: This one is in `locations/fixtures.py`, so it seems fine for it to be specific to locations.  I do agree though, that we should have a cached function which returns the `last_modified` for all of these, then this function can be rewritten to use that lookup.

File: corehq/apps/locations/tests/test_location_fixtures.py [hunk 7]
```diff
@@ -1,28 +1,97 @@
...
+        self.user = CommCareUser.create(self.domain, self.username, password)
+        self.user.save()
 
+    def tearDown(self):
+        all_users = CouchUser.all()
+        for user in all_users:
+            user.delete()
+        SQLLocation.objects.all().delete()
+        LocationType.objects.all().delete()
+        UserFixtureStatus.objects.all().delete()
```
Comment (esoergel) [inline corehq/apps/locations/tests/test_location_fixtures.py]: You might want to use [`delete_all_locations()`](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/locations/tests/util.py#L17-27) for the locations teardown - a bunch of random other crap gets created sometimes.  Also, it looks like you're not making an actual domain here, so it won't help you, but check out [`Domain.delete()`](https://github.com/dimagi/commcare-hq/blob/feb96ff3cfddf646e67fed42c7e3ed6d7cbd6641/corehq/apps/domain/models.py#L980-1000) if you're not already familiar - it does a cascading delete of a lot of things.

File: corehq/apps/locations/tests/test_location_fixtures.py [hunk 8]
```diff
@@ -1,28 +1,97 @@
...
+        LocationType.objects.all().update(last_modified=day_before_yesterday)
+        self.location_type = LocationType.objects.last()
+
+        location = SQLLocation(
+            location_id="unique-id",
+            domain="test-domain",
+            name="Meereen",
+            location_type=self.location_type,
+            metadata={'queen': "Daenerys Targaryen",
+                      'rebels': "Sons of the Harpy"},
```
Comment (esoergel) [inline corehq/apps/locations/tests/test_location_fixtures.py]: :imp: :hocho: :fire: :dragon_face: :princess:

File: corehq/apps/locations/tests/test_location_fixtures.py [hunk 9]
```diff
@@ -1,28 +1,97 @@
...
         }
         self.assertEquals(location_data, location.metadata)
+
+    def test_should_sync_locations_change_location_type(self):
+        """
+        When location_type gets changed, we should resync locations
+        Had to force update last_modified because it has the auto_now
+        property set
+        """
+        self.location_type.save()
```
Comment (esoergel) [inline corehq/apps/locations/tests/test_location_fixtures.py]: curious - why is this here rather than in `setUp`?

File: corehq/apps/users/models.py [hunk 10]
```diff
@@ -1766,12 +1766,25 @@ def sql_location(self):
...
+        from corehq.apps.fixtures.models import UserFixtureStatus
+        now = datetime.utcnow()
+        user_fixture_sync, new = UserFixtureStatus.objects.get_or_create(
+            user_id=self._id,
+            fixture_type=fixture_type,
+            defaults={'last_modified': now},
+        )
+        if not new:
+            user_fixture_sync.last_modified = now
+            user_fixture_sync.save()
```
Comment (esoergel) [inline corehq/apps/users/models.py]: This is a nice abstraction - I could see it being a public method too, though.

File: corehq/apps/locations/tests/test_location_fixtures.py (hunk 8 above)
Comment (czue) [inline corehq/apps/locations/tests/test_location_fixtures.py]: a test of :snowflake: and :fire:

Comment (biyeun) [inline corehq/apps/locations/tests/test_location_fixtures.py]: @_@ *sigh*

File: corehq/apps/locations/models.py (hunk 1 above)
Comment (dannyroberts) [inline corehq/apps/locations/models.py]: I actually can't say from my own experience because I just haven't used this before; but @czue's comments make sense and I have seen this used elsewhere like he said. My read is that this won't cause any timezone bugs, but only because our django project timezone (what a bizarre concept) is UTC.

**Comment by dimagimon** [comment]
There are 6 errors:

* corehq/apps/locations/fixtures.py, line 1 - F401 'datetime' imported but unused
* corehq/apps/locations/tests/test_location_fixtures.py, line 7 - F401 'fixture_last_modified' imported but unused
* corehq/apps/locations/tests/test_location_fixtures.py, line 8 - F401 'UserFixtureType' imported but unused
* corehq/apps/users/tests/fixture_status.py, line 3 - F401 'patch' imported but unused
* corehq/apps/users/tests/update.py, line 5 - F401 'UserFixtureStatus' imported but unused
* corehq/apps/users/tests/update.py, line 5 - F401 'UserFixtureType' imported but unused

File: corehq/apps/users/tests/fixture_status.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
+from datetime import datetime
+from django.test import TestCase
+from mock import MagicMock, patch
```
Comment (dimagimon) [inline corehq/apps/users/tests/fixture_status.py]: F401 'patch' imported but unused

File: corehq/apps/users/tests/fixture_status.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+            user_id=self.couch_user._id,
+            fixture_type=UserFixtureType.CHOICES[0][0],
+            last_modified=now,
+        ).save()
+
+        self.assertEqual(self.couch_user.fixture_status("fake_status"), UserFixtureStatus.DEFAULT_LAST_MODIFIED)
+        self.assertEqual(couch_user.fixture_status(UserFixtureType.CHOICES[0][0]), now)
+
+    def test_update_status_set_location(self):
+        fake_location = MagicMock()
```
Comment (dimagimon) [inline corehq/apps/users/tests/fixture_status.py]: F821 undefined name 'MagicMock'

File: corehq/apps/users/tests/fixture_status.py [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+        self.couch_user.set_location(fake_location)
+
+        self.assertEqual(UserFixtureStatus.objects.all().count(), 1)
+        user_fixture_status = UserFixtureStatus.objects.first()
+
+        self.assertEqual(user_fixture_status.user_id, self.couch_user._id)
+        self.assertEqual(user_fixture_status.fixture_type, UserFixtureType.LOCATION)
+
+    def test_update_status_unset_location(self):
+        fake_location = MagicMock()
```
Comment (dimagimon) [inline corehq/apps/users/tests/fixture_status.py]: F821 undefined name 'MagicMock'

File: corehq/apps/users/tests/fixture_status.py [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        self.couch_user.set_location(fake_location)
+        previously_updated_time = UserFixtureStatus.objects.get(user_id=self.couch_user._id).last_modified
+
+        self.couch_user.unset_location()
+
+        new_updated_time = UserFixtureStatus.objects.get(user_id=self.couch_user._id).last_modified
+        self.assertTrue(new_updated_time > previously_updated_time)
+
+    def test_update_status_reset_location(self):
+        fake_location = MagicMock()
```
Comment (dimagimon) [inline corehq/apps/users/tests/fixture_status.py]: F821 undefined name 'MagicMock'

File: corehq/apps/locations/fixtures.py (hunk 6 above)
Comment (proteusvacuum) [inline corehq/apps/locations/fixtures.py]: Changed this so that there is now a cached method on `user` which fetches the fixture status that is desired

File: corehq/apps/users/models.py (hunk 10 above)
Comment (proteusvacuum) [inline corehq/apps/users/models.py]: :+1:

**Comment by czue** [comment]
build fail looks real still:
```
======================================================================
FAIL: test_update_status_set_location (corehq.apps.users.tests.fixture_status.TestFixtureStatus)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/users/tests/fixture_status.py", line 56, in test_update_status_set_location
    self.assertEqual(UserFixtureStatus.objects.all().count(), 0)
AssertionError: 3 != 0
```

**Comment by proteusvacuum** [comment]
:white_check_mark:

**Comment by proteusvacuum** [comment]
@czue @snopoke @esoergel integrated a bunch of your suggestions.

File: corehq/apps/users/models.py [hunk 15]
```diff
@@ -2001,6 +2007,39 @@ def get_location_map_case(self):
         except ResourceNotFound:
             return None
 
+    @property
+    @skippable_quickcache([], lambda _: settings.UNIT_TESTING)
```
Comment (czue) [inline corehq/apps/users/models.py]: this is a neat pattern

Comment (czue) [inline corehq/apps/users/models.py]: does the `quickcache` `vary_on` automatically include `self`? I don't think I've seen it used like this before and am worried this is actually caching a singleton when you want an instance level cache.
cc @dannyroberts

**Comment by czue** [comment]
one possible blocker. otherwise :+1: to merge from me. assume you'll remove the label when you're ready as well.

File: corehq/apps/users/models.py (hunk 15 above)
Comment (snopoke) [inline corehq/apps/users/models.py]: Correct, this will function as a singleton cache. You should add `'self._id'` as a vary on parameter.

**Comment by snopoke** [comment]
:+1: from me once quickcache is fixed
